### PR TITLE
LR Changes to Support Uncompressed Max Write Size Change

### PR DESCRIPTION
**Pull request by pankti-m**
Description:
PR [3828](https://github.com/CorfuDB/CorfuDB/pull/3828) introduces a limit of 100MB on the max uncompressed write size for a transaction.  Currently, Log Replicator receiver applies every batch of incoming Snapshot Sync data in a single transaction in the transfer phase.

Currently, LR's batch size for transfer is limited to 64MB which is well within the 100MB limit.  But it is customizable, so the limit of (100MB + some buffer for LR metadata and serialization) must be enforced to allow the replicated data to be applied atomically on the receiver.  This uncompressed transaction size must also be used when chunking the aggregated data in the apply phase.

Why should this be merged:  Enforces Corfu Runtime's max write size in LR

Related issue(s) (if applicable): #<number>

## Checklist (Definition of Done):

- [ ] There are no TODOs left in the code
- [ ] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [ ] Change is covered by automated tests
- [ ] Public API has Javadoc

### CODE REVIEW

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/StreamsSnapshotWriter.java [hunk 1]
```diff
@@ -155,9 +155,34 @@ private void processUpdatesShadowStream(List<SMREntry> smrEntries, Long currentS
                                             UUID snapshotSyncId) {
         CorfuStoreMetadata.Timestamp timestamp;
 
+        int size = 0;
+        for (SMREntry smrEntry : smrEntries) {
+            size += smrEntry.getSerializedSize();
+        }
+        log.info("Total Size of Data = {}", size);
```
Comment (shama358) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/StreamsSnapshotWriter.java:162]: Are we planning on checking the size on the receiver side even in the final cut?

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuInterClusterReplicationServer.java [hunk 2]
```diff
@@ -377,7 +377,7 @@ private CorfuRuntime getRuntime(ServerContext serverContext) {
             .tlsEnabled((Boolean) serverContext.getServerConfig().get("--enable-tls"))
             .systemDownHandler(() -> System.exit(SYSTEM_EXIT_ERROR_CODE))
             .maxCacheEntries(serverContext.getLogReplicationCacheMaxSize()/2)
-            .maxWriteSize(serverContext.getMaxWriteSize())
+            .maxUncompressedWriteSize(serverContext.getMaxWriteSize())
```
Comment (shama358) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuInterClusterReplicationServer.java:380]: I think we should have a check to make sure we don't go over the default maxUncompressedWriteSize.

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/config/LogReplicationConfig.java [hunk 3]
```diff
@@ -103,6 +103,6 @@ public LogReplicationConfig(@NonNull LogReplicationSession session,
             this.maxCacheSize = serverContext.getLogReplicationCacheMaxSize();
             this.maxSnapshotEntriesApplied = serverContext.getMaxSnapshotEntriesApplied();
         }
-        this.maxDataSizePerMsg = maxMsgSize * DATA_FRACTION_PER_MSG / 100;
+        this.maxDataSizePerMsg = Math.min(maxMsgSize, serverContext.getMaxWriteSize() * DATA_FRACTION_PER_MSG / 100);
```
Comment (shama358) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/config/LogReplicationConfig.java:106]: I think we should have a check to make sure we are atleast 80% of the default maxUncompressedWriteSize. Just so we minimize possibility of going over

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/StreamsSnapshotWriter.java (hunk 1 above)
Comment (pankti-m) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/StreamsSnapshotWriter.java:162]: No not checking.  But we will have to keep a buffer because final serialization before commit can add some bytes to what we compute.  The above was just for debugging and I will remove it.

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/StreamsSnapshotWriter.java [hunk 4]
```diff
@@ -348,8 +348,9 @@ private void applyShadowStream(UUID streamId, long snapshot) {
             // OOM on applications running with a small memory footprint.  So for such tables, introduce an
             // additional limit of max number of entries(50 by default) applied in a single transaction.  This
             // algorithm is in line with the limits imposed in Compaction and Restore workflows.
-            if (bufferSize + smrEntry.getSerializedSize() > metadataManager.getRuntime().getParameters()
-                    .getMaxWriteSize() || maxEntriesLimitReached(streamId, buffer)) {
+            if (bufferSize + smrEntry.getSerializedSize() > replicationContext.getConfig(session).getMaxTransferSize()
```
[the remark this answers is not in the document]
Comment (pankti-m) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/StreamsSnapshotWriter.java:351]: True, but we have to keep a buffer for this check also.  I had noticed in our tests that the final write size can increase beyond the runtime's limit despite having passed this check.  We can keep a lower buffer than the transfer size though.  Let me know what you think.

[the remark this answers is not in the document]
Comment (shama358) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/StreamsSnapshotWriter.java:351]: I see...would the same 80% of uncompressed size work here as well?
Since the transfer size is configurable, there is no guarantee of having it as 80% of the size..

[the remark this answers is not in the document]
Comment (pankti-m) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/StreamsSnapshotWriter.java:351]: Yes, I added a new variable for the apply size which considers just a fraction(85%) of the max uncompressed runtime size.  This will make it independent of the transfer size and allow us to write more data in a single transaction in the apply phase.

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/logreader/BaseSnapshotReader.java [hunk 5]
```diff
@@ -141,24 +141,23 @@ private SMREntryList next(OpaqueStreamIterator stream) {
...
 
         try {
-            while (currentMsgSize < maxDataSizePerMsg) {
+            while (currentMsgSize < maxTransferSize) {
                 if (lastEntry != null) {
                     List<SMREntry> smrEntries = lastEntry.getEntries().get(stream.uuid);
                     if (smrEntries != null) {
                         int currentEntrySize = ReaderUtility.calculateSize(smrEntries);
-
                         if (currentEntrySize > DEFAULT_MAX_DATA_MSG_SIZE) {
```
Comment (shama358) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/logreader/BaseSnapshotReader.java:149]: We should enhance this check now?
This check now makes sense only if maxTransferSize > DEFAULT_MAX_DATA_MSG_SIZE. Otherwise LR will only support moving records which are at the max `maxTransferSize`

Comment (pankti-m) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/logreader/BaseSnapshotReader.java:149]: The recommended size limit for a protobuf message is 64MB.  Here we are checking against that and failing the read if the size is beyond 64MB.  On L153, the size is compared against the transfer size.  I think both checks are valid?  Let me know if I did not follow your comment fully...

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/StreamsSnapshotWriter.java [hunk 6]
```diff
@@ -157,7 +157,12 @@ private void processUpdatesShadowStream(List<SMREntry> smrEntries, Long currentS
 
         try (TxnContext txn = metadataManager.getTxnContext()) {
             updateLog(txn, smrEntries, shadowStreamUuid);
-            metadataManager.updateReplicationMetadataField(txn, session, ReplicationMetadata.LASTSNAPSHOTTRANSFERREDSEQNUMBER_FIELD_NUMBER, currentSeqNum);
+            ReplicationMetadata metadata = metadataManager.queryReplicationMetadata(txn, session);
+            ReplicationMetadata updatedMetadata = metadata.toBuilder().setLastSnapshotStarted(srcGlobalSnapshot)
```
Comment (hisundar) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/StreamsSnapshotWriter.java:161]: Just curious - are these changes needed for the uncompressed size handling or is it a separate change?

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/config/LogReplicationConfig.java [hunk 7]
```diff
@@ -60,16 +61,18 @@ public abstract class LogReplicationConfig {
...
-     * The max size of data payload for the log replication message.
+     * The max size of replicated data payload transferred at a time.
+     */
+    private int maxTransferSize;
+
+    /**
+     * The max size of data payload written in a single transaction during the Apply phase of Snapshot Sync
      */
-    private int maxDataSizePerMsg;
+    private int maxApplySize;
```
Comment (hisundar) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/config/LogReplicationConfig.java:75]: can we please make this a long to avoid the overflow just in case?

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/config/LogReplicationConfig.java [hunk 8]
```diff
@@ -60,16 +61,18 @@ public abstract class LogReplicationConfig {
...
-    private int maxMsgSize;
-
     // Max Cache number of entries
     private int maxCacheSize;
 
     /**
-     * The max size of data payload for the log replication message.
+     * The max size of replicated data payload transferred at a time.
+     */
+    private int maxTransferSize;
```
Comment (hisundar) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/config/LogReplicationConfig.java:70]: can we please make this a long to avoid any computational overflows accidentally?

**Review by hisundar — COMMENTED** [review]
Looks good just a couple of questions/comments. thanks

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/StreamsSnapshotWriter.java (hunk 6 above)
Comment (pankti-m) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/StreamsSnapshotWriter.java:161]: they are indirectly needed..

Earlier, we wrote the updated metadata multiple times in a given transaction, once for each field(topologyId, last snapshot started and last transferred).  This increased the transaction size and was not required as all fields are in the same message.  I have consolidated it in a single update.

Let me know if you have any further questions.

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/logreader/BaseSnapshotReader.java (hunk 5 above)
Comment (shama358) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/logreader/BaseSnapshotReader.java:149]: if the data size is less than 64MB but say more than maxTransferSize, we will be stuck in this loop forever. I am thinking about throwing an exception and maybe putting that particular session in the STOPPED state. That way other sessions will not be affected (more so with the thread pool change. Less starvation)

Comment (pankti-m) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/logreader/BaseSnapshotReader.java:149]: As of now, the plan is to never expose the runtime uncompressed tx size so it will be 100MB and always more than the max data msg size(64 MB).  

But in future if it changes, this is a valid point.  So added the fix.

Also, the ideal fix here should be to split the transaction as currentEntrySize is calculated for all smr entries in the transaction.  As it will be a basic functional change, I am not adding it now and just put a TODO for it.

**Review by shama358 — APPROVED** [review]
LGTM. The PR can be merged after reverting the last commit (11e9929b96475e9a719ee49c4e9978cf9a5f092a)

**Review by hisundar — APPROVED** [review]
lgtm
